src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,1087 @@
+        // TODO: Retrieve the tools from the Azure Key Vault, when passed all tests.
+        NoSeriesCopilotSetup.Get();
+        exit(NoSeriesCopilotSetup.GetTool2OutputFormatPromptFromIsolatedStorage())
+    end;
+
+    [NonDebuggable]
+    internal procedure GenerateNoSeries(SystemPromptTxt: SecretText; InputText: Text): Text
+    var
+        AzureOpenAI: Codeunit "Azure OpenAi";
+        AOAIDeployments: Codeunit "AOAI Deployments";
+        AOAIOperationResponse: Codeunit "AOAI Operation Response";
+        AOAIChatCompletionParams: Codeunit "AOAI Chat Completion Params";
+        AOAIChatMessages: Codeunit "AOAI Chat Messages";
+        CompletionAnswerTxt: Text;
+        ToolJson: JsonObject;
+    begin
+        if not AzureOpenAI.IsEnabled(Enum::"Copilot Capability"::"No. Series Copilot") then
+            exit;
+
+        AzureOpenAI.SetAuthorization(Enum::"AOAI Model Type"::"Chat Completions", GetEndpoint(), GetDeployment(), GetSecret());
+        AzureOpenAI.SetCopilotCapability(Enum::"Copilot Capability"::"No. Series Copilot");
+        AOAIChatCompletionParams.SetMaxTokens(MaxOutputTokens());
+        AOAIChatMessages.AddSystemMessage(SystemPromptTxt.Unwrap());
+
+        foreach ToolJson in GetTools() do
+            AOAIChatMessages.AddTool(ToolJson);
+
+        AOAIChatMessages.AddUserMessage(InputText);
+        AzureOpenAI.GenerateChatCompletion(AOAIChatMessages, AOAIChatCompletionParams, AOAIOperationResponse);
+        if AOAIOperationResponse.IsSuccess() then
+            CompletionAnswerTxt := AOAIChatMessages.GetLastMessage()
+        else
+            Error(AOAIOperationResponse.GetError());
+
+        if AOAIChatMessages.IsToolsList(CompletionAnswerTxt) then
+            CompletionAnswerTxt := CallTool(AzureOpenAI, AOAIChatMessages, AOAIChatCompletionParams, CompletionAnswerTxt);
+
+        exit(CompletionAnswerTxt);
+    end;
+
+    local procedure CallTool(var AzureOpenAI: Codeunit "Azure OpenAi"; var AOAIChatMessages: Codeunit "AOAI Chat Messages"; var AOAIChatCompletionParams: Codeunit "AOAI Chat Completion Params"; var ToolDefinition: Text): Text
+    var
+        ToolCallId: Text;
+        FunctionName, FunctionArguments, ToolResponse : Text;
+        ToolResponses: Dictionary of [Text, Integer]; // tool response can be a list of strings, as the response can be too long and exceed the token limit. In this case each string would be a separate message, each of them should be called separately. The integer is the number of tables used in the prompt, so we can test if the LLM answer covers all tables
+        FinalResults: List of [Text]; // The final response will be the concatenation of all the LLM responses (final results).        MaxToolResponseTokenLength, i : Integer;
+        MaxToolResponseTokenLength, ExpectedNoSeriesCount, i : Integer;
+    begin
+        // remove the tools from the chat messages, as they are not needed anymore
+        for i := 1 to AOAIChatMessages.GetTools().Count do
```

Comment:
https://github.com/microsoft/BCApps/pull/824
The above PR will allow you to remove all the tools more easily.
`AOAIChatMessages.ClearTools();`